File: app-flutter/lib/core/error/failures.dart
import 'package:equatable/equatable.dart';

/// Base Failure class
/// All failures in the domain layer extend this
abstract class Failure extends Equatable {
  final String message;
  final int? statusCode;

  const Failure(this.message, [this.statusCode]);

  @override
  List<Object?> get props => [message, statusCode];
}

/// Network-related failures
class NetworkFailure extends Failure {
  const NetworkFailure([super.message = 'Network error occurred']);
}

/// Server-related failures (5xx errors)
class ServerFailure extends Failure {
  const ServerFailure([
    super.message = 'Server error occurred',
    super.statusCode,
  ]);
}

/// Client-related failures (4xx errors)
class ClientFailure extends Failure {
  const ClientFailure([
    super.message = 'Client error occurred',
    super.statusCode,
  ]);
}

/// Timeout failures
class TimeoutFailure extends Failure {
  const TimeoutFailure([super.message = 'Request timeout']);
}

/// Cancellation failures
class CancellationFailure extends Failure {
  const CancellationFailure([super.message = 'Request cancelled']);
}

/// Cache failures
class CacheFailure extends Failure {
  const CacheFailure([super.message = 'Cache error occurred']);
}

/// Unknown failures
class UnknownFailure extends Failure {
  const UnknownFailure([super.message = 'An unknown error occurred']);
}

/// Validation failures
class ValidationFailure extends Failure {
  const ValidationFailure([super.message = 'Validation error']);
}

/// Not found failures (404)
class NotFoundFailure extends Failure {
  const NotFoundFailure([String message = 'Resource not found'])
    : super(message, 404);
}

/// Unauthorized failures (401)
class UnauthorizedFailure extends Failure {
  const UnauthorizedFailure([String message = 'Unauthorized access'])
    : super(message, 401);
}

/// Forbidden failures (403)
class ForbiddenFailure extends Failure {
  const ForbiddenFailure([String message = 'Access forbidden'])
    : super(message, 403);
}

File: nebras-app/app/src/main/kotlin/com/nebras/mobile/core/error/Failures.kt
package com.nebras.mobile.core.error

/**
 * الصفّ الأساس لكلّ الإخفاقات في طبقة الـ domain.
 * مقابل `core/error/failures.dart` (كان يعتمد `Equatable` — `data class`
 * في Kotlin يوفّر المساواة البنيويّة نفسها مجّاناً).
 */
sealed class Failure(
    open val message: String,
    open val statusCode: Int? = null,
)

data class NetworkFailure(
    override val message: String = "Network error occurred",
) : Failure(message)

data class ServerFailure(
    override val message: String = "Server error occurred",
    override val statusCode: Int? = null,
) : Failure(message, statusCode)

data class ClientFailure(
    override val message: String = "Client error occurred",
    override val statusCode: Int? = null,
) : Failure(message, statusCode)

data class TimeoutFailure(
    override val message: String = "Request timeout",
) : Failure(message)

data class CancellationFailure(
    override val message: String = "Request cancelled",
) : Failure(message)

data class CacheFailure(
    override val message: String = "Cache error occurred",
) : Failure(message)

data class UnknownFailure(
    override val message: String = "An unknown error occurred",
) : Failure(message)

data class ValidationFailure(
    override val message: String = "Validation error",
) : Failure(message)

data class NotFoundFailure(
    override val message: String = "Resource not found",
) : Failure(message, 404)

data class UnauthorizedFailure(
    override val message: String = "Unauthorized access",
) : Failure(message, 401)

data class ForbiddenFailure(
    override val message: String = "Access forbidden",
) : Failure(message, 403)
